### ISSUE TITLE
Stop when a library is mixed FR and RF orientation

### ISSUE THREAD

**Issue by sjackman** [bcgsc/abyss#164]
When a library is mixed FR and RF orientation, ABySS currently chooses whichever orientation is dominant and plows ahead. Nextera mate-pair libraries can have more FR content than RF content. See below for an example, which has 58% FR. Quietly choosing FR for the library orientation in this case is wrong in this case. It does output the message `The library mp1-6.hist is oriented forward-reverse (FR).` if verbose `-v` is specified, but without `-v` it prints nothing. The resulting assembly has terrible contiguity and is full of misassemblies. With a reference, it's not readily apparent why the assembly is bad. Without a reference, the user may have no idea that the assembly is bad.

If the library is FR (>99% FR) it should quietly go ahead with FR.
If the library is RF (>98% RF) it should quietly go ahead with RF.
Any intermediate value should exit with failure and print an error message suggesting that the user run a tool, such as NxTrim, to trim the Nextera mate-pair adapters. If the user manually specifies `--fr` or `--rf` it should still print the warning, but continue without an error.

**Comment by sjackman** [comment]
Here's an example data set and assembly:
https://github.com/sjackman/abyss-dmelanogaster


```
Mapped 50458909 of 58272472 reads (86.6%)
Mapped 50198869 of 58272472 reads uniquely (86.1%)
Read 58272472 alignments
Mateless          0
Unaligned   1863247  6.39%
Singleton   4087069  14%
FR          7418260  25.5%
RF          5372299  18.4%
FF           323713  1.11%
Different  10071648  34.6%
Total      29136236
FR Stats mean: 313.8 median: 297 sd: 118.7 n: 7405162 min: 3 max: 2469 ignored: 5385397
   _▁▃▆███▇▅▄▃▂▁__
RF Stats mean: 8661 median: 9039 sd: 2023 n: 5371423 min: 2 max: 11886 ignored: 7419136
                                                   __▁▂▃▅▆███▇▇▆▅▄▄▃▃▂▁▁__
Mate orientation FR: 7418260 (58%) RF: 5372299 (42%)
The library mp1-6.hist is oriented forward-reverse (FR).
Stats mean: 313.8 median: 297 sd: 118.7 n: 7405162 min: 3 max: 2469
   _▁▃▆███▇▅▄▃▂▁__
Minimum and maximum distance are set to -31 and 2469 bp.
Duplicate rate of spanning fragments: 5715371/12814478 (44.6%)
```

**Comment by sjackman** [comment]
> If the library is FR (>99% FR) it should quietly go ahead with FR.
If the library is RF (>98% RF) it should quietly go ahead with RF.

@jaredo (author of NxTrim) suggested that >95% RF may be a better value.
https://github.com/sequencing/NxTrim/issues/50#issuecomment-327263192

**Comment by benvvalk** [comment]
Thank for troubleshooting this @sjackman and I agree with your recommendation to add a warning. (I'll add it to my TODO list).

**Comment by sjackman** [comment]
ABySS should print the orientation of the detected library whether or not `-v` is specified. These two lines:
```
Mate orientation FR: 7418260 (58%) RF: 5372299 (42%)
The library mp1-6.hist is oriented forward-reverse (FR).
```
If the selected orientation is <95% it should print a warning. If it's less than some threshold, I think it should be a hard error, suggesting that they run `NxTrim` or specify the desired orientation using `--rf` or `--fr`.

**Comment by benvvalk** [comment]
Noted and sounds good to me.

**Comment by sjackman** [comment]
Thoughts on what that threshold should be? A paired-end library should have <1% RF orientation. A mate-pair library can have ~50% RF orientation, and after NxTrim can have ~1% to ~5%. I'm thinking if the RF content is in the range of…

- 0 to 2%, good it's FR oriented
- 2% to 50%, print an error suggesting NxTrim `nxtrim --justmp --rf`, and exit
- 50% to 95%, print a warning suggesting NxTrim `nxtrim --justmp --rf`, and select RF oriented
- 95% to 100%, good it's RF oriented

Thoughts?

**Comment by benvvalk** [comment]
@sjackman Seems pretty reasonable to me.

**Comment by mmokrejs** [comment]
Why cannot a scaffolder work with the fact that there is a mis of two population of reads? The logs report some 'ignored' read pairs. Is that actually saying *if the dataset is to be treated as FR then than many read pairs will be ignored*? This appears there twice for RF and FR orientation. So where is the summary information printed in the log? Could those *ignored* reads be written into a FASTQ file for inspection or latter re-use as say FR pairs?

Would you please explain what is the meaning of `Mateless`, `Unaligned`, `Different`? The second column speaks about reads or read pairs? or k-mers or alignments? I could give it a more thought but I want to make you (decently) to improve the log file too. Thank you.

The last line saying `Total` should include 100% (a proper sum of the percentages above) at its end and re-state if that is about reads halves or read pairs or whichever.

Further it would be very helpful if abyss dumped out several text files listing the reads (or read pairs?) per each category. That would mean 7 TXT files per each input library file. I would like to inspect especially those FR vs. RF vs. `Different`. I wouldn't mind if it even dumped out interleaved FASTQ files, practically separating my input data into those categories. Would be helpfull for debugging and bug reporting.

First, here are my paired-end libs, both Illumina 2x250 from HiSeq2500:

Abyss used is 2.0.2 version.

```
abyss-map -v  -j104 -l40    /scratch/mygenome/tt_16D1C3L12/paired_end_619bp//HKMHTBCXX.1.tt_16D1C3L12.trimmomatic.paired.prinseq.minlen20.19552.19552.fastq tt_16D1C3L12-3.fa \
	|abyss-fixmate -v  -l40  -h HKMHTBCXX.1.pe-3.hist \
	|sort -snk3 -k4 \
	|DistanceEst -v  -j104 -k128 -l40 -s1000 -n10   -o HKMHTBCXX.1.pe-3.dist HKMHTBCXX.1.pe-3.hist
Reading from standard input...Reading `
tt_16D1C3L12-3.fa'...
Using 202 MB of memory and 52.6 B/sequence.
Reading `tt_16D1C3L12-3.fa'...
Building the suffix array...
Building the Burrows-Wheeler transform...
Building the character occurrence table...
Using 15.4 GB of memory and 8.9 B/bp.
Mapped 311699312 of 317650862 reads (98.1%)
Mapped 287389021 of 317650862 reads uniquely (90.5%)
Mateless      188974  0.119%
Unaligned     146244  0.092%
Singleton    5650518  3.56%
FR          70766637  44.5%
RF             36136  0.0227%
FF              1938  0.00122%
Different   82129471  51.7%
Total      158919918
FR Stats mean: 598.7 median: 592 sd: 126.4 n: 70759811 min: 11 max: 1294 ignored: 42962
                    __▁▁▂▃▄▅▅▆▇█████▇▇▆▅▄▄▃▂▂▁▁___
Mate orientation FR: 70766637 (99.9%) RF: 36136 (0.051%)
The library HKMHTBCXX.1.pe-3.hist is oriented forward-reverse (FR).
Stats mean: 598.7 median: 592 sd: 126.4 n: 70759811 min: 11 max: 1294
                    __▁▁▂▃▄▅▅▆▇█████▇▇▆▅▄▄▃▂▂▁▁___
Minimum and maximum distance are set to -127 and 1294 bp.
Duplicate rate of spanning fragments: 304081/12357355 (2.46%)
abyss-map -v  -j104 -l40    /scratch/mygenome/tt_16D1C3L12/paired_end_497bp_201709//HT5V3BCXY.2.tt_16D1C3L12.trimmomatic.paired.prinseq.minlen20.3091.3091.fastq tt_16D1C3L12-3.fa \
	|abyss-fixmate -v  -l40  -h HT5V3BCXY.2.pe-3.hist \
	|sort -snk3 -k4 \
	|DistanceEst -v  -j104 -k128 -l40 -s1000 -n10   -o HT5V3BCXY.2.pe-3.dist HT5V3BCXY.2.pe-3.hist
Reading from standard input...
Reading `tt_16D1C3L12-3.fa'...
Using 202 MB of memory and 52.6 B/sequence.
Reading `tt_16D1C3L12-3.fa'...
Building the suffix array...
Building the Burrows-Wheeler transform...
Building the character occurrence table...
Using 15.4 GB of memory and 8.9 B/bp.
Mapped 347486688 of 359709190 reads (96.6%)
Mapped 317997397 of 359709190 reads uniquely (88.4%)
Mateless      248664  0.138%
Unaligned     270501  0.15%
Singleton   11666557  6.48%
FR          79745851  44.3%
RF             40543  0.0225%
FF              7656  0.00425%
Different   87999155  48.9%
Total      179978927
FR Stats mean: 584.3 median: 596 sd: 177.5 n: 79738120 min: 11 max: 1298 ignored: 48274
             _▃▃▃▃▃▃▃▄▄▄▄▄▅▅▆▆▆▇▇████████▇▇▆▅▄▄▃▂▂▁▁___
Mate orientation FR: 79745851 (99.9%) RF: 40543 (0.0508%)
The library HT5V3BCXY.2.pe-3.hist is oriented forward-reverse (FR).
Stats mean: 584.3 median: 596 sd: 177.5 n: 79738120 min: 11 max: 1298
             _▃▃▃▃▃▃▃▄▄▄▄▄▅▅▆▆▆▇▇████████▇▇▆▅▄▄▃▂▂▁▁___
Minimum and maximum distance are set to -127 and 1298 bp.
Duplicate rate of spanning fragments: 801992/14129681 (5.68%)
```

[paired-ends.log](https://github.com/bcgsc/abyss/files/1702981/paired-ends.log)



Here are three mate-pair libraries, the 5kbp and 8kbp from year 2016 were 2x150 but one 5kbp from 2017 were sequenced as 2x250. Linker removal was done using splitnextera.sh from BBmap bundle because NxTrim and nextclip are just not removing linkers properly. I do not recommend anybody to use these tools. You can find on github my discussion with author of nxtrim who was very helpful but in the end he admitted the would be much more work need to add proper smith-waterman alignment and scoring. without both it will not work.

```
BBmap-37.67/bbduk.sh in=$p.fastq out=$p.bbduk.replaced.fq kmask=J k=13 editdistance=2 mink=8 editdistance2=2 literal=CTGTCTCTTATACACATCTAGATGTGTATAAGAGACAG -da || exit 255

BBmap-37.67/splitnextera.sh in=$p.bbduk.replaced.fq out=$p.bbduk.splitnextera.lmp.fq outf=$p.bbduk.splitnextera.fragments.fq outu=$p.bbduk.splitnextera.unknown.fq outs=$p.bbduk.splitnextera.singletons.fq || exit 255
```

Below I list only those *.lmp.* file data, which should be read pairs which had a linker at all and which was removed. So those should be true LMP pairs. The *.fragments.* are to various degree paired-end read contaminants and the *.unknown.* is likewise, to varying degree RF and FR mix (mostly? more RF pairs so I would like to include them during scaffolding).

```
abyss-map -v  -j104 -l40    /scratch/mygenome/tt_16D1C3L12/mate_pairs_201609//HFYJ5AFXX.5kb.trimmomatic.bbduk.splitnextera.lmp.fq tt_16D1C3L12-6.fa \
	|abyss-fixmate -v  -l40  -h HFYJ5AFXX.5kb.lmp-6.hist \
	|sort -snk3 -k4 \
	|DistanceEst --dot --mean -v  -j104 -k128 -l40 -s1000 -n10   -o HFYJ5AFXX.5kb.lmp-6.dist.dot HFYJ5AFXX.5kb.lmp-6.hist
Reading from standard input...Reading `
tt_16D1C3L12-6.fa'...
Using 202 MB of memory and 67 B/sequence.
Reading `tt_16D1C3L12-6.fa'...
Building the suffix array...
Building the Burrows-Wheeler transform...
Building the character occurrence table...
Using 14.6 GB of memory and 8.9 B/bp.
Mapped 59455198 of 63241054 reads (94%)
Mapped 43375465 of 63241054 reads uniquely (68.6%)
Mateless     953342  2.97%
Unaligned    579194  1.8%
Singleton   2578619  8.03%
FR            13540  0.0422%
RF          5448089  17%
FF            65020  0.203%
Different  22459394  70%
Total      32097198
FR Stats mean: 487.3 median: 212 sd: 730.1 n: 12744 min: 0 max: 8206 ignored: 5448885
█▄▃▁_
RF Stats mean: 4013 median: 4029 sd: 679.9 n: 5447122 min: 6 max: 7003 ignored: 14507
                                    __▂▄▆█████▇▇▆▆▅▅▄▄▃▂▂▁_
Mate orientation FR: 13540 (0.248%) RF: 5448089 (99.8%)
The library HFYJ5AFXX.5kb.lmp-6.hist is oriented reverse-forward (RF).
Stats mean: 4013 median: 4029 sd: 679.9 n: 5447122 min: 6 max: 7003
                                    __▂▄▆█████▇▇▆▆▅▅▄▄▃▂▂▁_
Minimum and maximum distance are set to -127 and 7003 bp.
Duplicate rate of spanning fragments: 243007/8320353 (2.92%)
abyss-map -v  -j104 -l40    /scratch/mygenome/tt_16D1C3L12/mate_pairs_201609//HFYJ5AFXX.8kb.trimmomatic.bbduk.splitnextera.lmp.fq tt_16D1C3L12-6.fa \
	|abyss-fixmate -v  -l40  -h HFYJ5AFXX.8kb.lmp-6.hist \
	|sort -snk3 -k4 \
	|DistanceEst --dot --mean -v  -j104 -k128 -l40 -s1000 -n10   -o HFYJ5AFXX.8kb.lmp-6.dist.dot HFYJ5AFXX.8kb.lmp-6.hist
Reading `Reading from standard input...tt_16D1C3L12-6.fa
'...
Using 202 MB of memory and 67 B/sequence.
Reading `tt_16D1C3L12-6.fa'...
Building the suffix array...
Building the Burrows-Wheeler transform...
Building the character occurrence table...
Using 14.6 GB of memory and 8.9 B/bp.
Mapped 58596424 of 62369102 reads (94%)
Mapped 42660735 of 62369102 reads uniquely (68.4%)
Mateless    1033464  3.26%
Unaligned    557796  1.76%
Singleton   2600213  8.2%
FR            19181  0.0605%
RF          3231051  10.2%
FF           244490  0.771%
Different  24015088  75.8%
Total      31701283
FR Stats mean: 445.2 median: 279 sd: 576.9 n: 18249 min: 0 max: 5713 ignored: 3231983
▆▅▅██▃__
RF Stats mean: 5970 median: 6069 sd: 2156 n: 3230107 min: 2 max: 16553 ignored: 20125
  _____________▁▁___▁▁▁▂▄▇███▇▇▆▆▆▅▄▃▃▂▂▁▁_____
Mate orientation FR: 19181 (0.59%) RF: 3231051 (99.4%)
The library HFYJ5AFXX.8kb.lmp-6.hist is oriented reverse-forward (RF).
Stats mean: 5970 median: 6069 sd: 2156 n: 3230107 min: 2 max: 16553
  _____________▁▁___▁▁▁▂▄▇███▇▇▆▆▆▅▄▃▃▂▂▁▁_____
Minimum and maximum distance are set to -127 and 16553 bp.
Duplicate rate of spanning fragments: 264302/9819646 (2.69%)
abyss-map -v  -j104 -l40    /scratch/mygenome/tt_16D1C3L12/mate_pairs_201709//HWFNLBCXY.2.tt_16D1C3L12.trimmomatic.bbduk.splitnextera.lmp.fq tt_16D1C3L12-6.fa \
	|abyss-fixmate -v  -l40  -h HWFNLBCXY.lmp-6.hist \
	|sort -snk3 -k4 \
	|DistanceEst --dot --mean -v  -j104 -k128 -l40 -s1000 -n10   -o HWFNLBCXY.lmp-6.dist.dot HWFNLBCXY.lmp-6.hist
Reading from standard input...Reading `
tt_16D1C3L12-6.fa'...
Using 202 MB of memory and 67 B/sequence.
Reading `tt_16D1C3L12-6.fa'...
Building the suffix array...
Building the Burrows-Wheeler transform...
Building the character occurrence table...
Using 14.6 GB of memory and 8.9 B/bp.
Mapped 41756673 of 42620602 reads (98%)
Mapped 32273294 of 42620602 reads uniquely (75.7%)
Mateless     835594  3.85%
Unaligned     37081  0.171%
Singleton    770627  3.55%
FR            22793  0.105%
RF          3564129  16.4%
FF            96922  0.446%
Different  16400952  75.5%
Total      21728098
FR Stats mean: 305.9 median: 178 sd: 431.7 n: 22018 min: 0 max: 4398 ignored: 3564904
▃█▇▄▃▃▁__
RF Stats mean: 4169 median: 4236 sd: 1162 n: 3563617 min: 2 max: 7642 ignored: 23305
                   _________________▁▂▃▆████▇▆▆▅▄▄▄▃▃▂▂▂▁▁▁▁_____
Mate orientation FR: 22793 (0.635%) RF: 3564129 (99.4%)
The library HWFNLBCXY.lmp-6.hist is oriented reverse-forward (RF).
Stats mean: 4169 median: 4236 sd: 1162 n: 3563617 min: 2 max: 7642
                   _________________▁▂▃▆████▇▆▆▅▄▄▄▃▃▂▂▂▁▁▁▁_____
Minimum and maximum distance are set to -127 and 7642 bp.
Duplicate rate of spanning fragments: 202311/5830779 (3.47%)
```
[scaffolding.log](https://github.com/bcgsc/abyss/files/1702961/scaffolding.log)

**Comment by sjackman** [comment]
> Would you please explain what is the meaning of Mateless, Singleton, Unaligned, Different?

- Mateless: no matching mate in the reads file.
- Singleton: One read aligned, one unaligned
- Unaligned: Both reads unaligned
- Different: Each read maps to a different contig. This category is used for scaffolding.

> The second column speaks about reads or read pairs? or k-mers or alignments? I could give it a more thought but I want to make you (decently) to improve the log file too. Thank you. The last line saying Total should include 100% (a proper sum of the percentages above) at its end and re-state if that is about reads halves or read pairs or whichever.

Read pairs, except mateless, which are single reads.

> Further it would be very helpful if abyss dumped out several text files listing the reads (or read pairs?) per each category. That would mean 7 TXT files per each input library file. I would like to inspect especially those FR vs. RF vs. Different.

I would also like ABySS to output these stats to TSV files. I'm afraid it's not a priority as we're limited for developer time. You can scrape them out of the log files. Alternatively, it's pretty straight forward to calculate these metrics from the BAM file of reads mapped to the contigs using `samtools flagstat`.

> I wouldn't mind if it even dumped out interleaved FASTQ files, practically separating my input data into those categories. Would be helpfull for debugging and bug reporting.

Add to your command line `abyss-pe pe-bam` or `abyss-pe mp-bam` for the contig and scaffold stages respectively, or if you prefer `pe-sam` and `mp-sam` for `sam.gz` files rather than BAM files.

**Comment by sjackman** [comment]
> Why cannot a scaffolder work with the fact that there is a mix of two populations of reads?

If you have Nextera mate-pair libraries, you can use NxTrim to split your reads into RF and FR sets, and give ABySS both those libraries. We don't usually use the FR reads, but it's possible that they could help the assembly.

> The logs report some 'ignored' read pairs. Is that actually saying If the dataset is to be treated as FR then that many read pairs will be ignored? This message appears there twice for RF and FR orientation. So where is the summary information printed in the log? Could those ignored reads be written into a FASTQ file for inspection or latter re-use as say FR pairs?

`trimmed` would be a better term. The distribution of fragment sizes calculated from read pairs that map to the same contig is first trimmed (0.00005 from each end) before calculating the mean and standard deviation. Read pairs that map to the same contig are not useful for scaffolding, other than for this purpose of calculating the empirical distribution of fragment sizes.

**Comment by mmokrejs** [comment]
> If you have Nextera mate-pair libraries, you can use NxTrim to split your reads into RF and FR sets, and give ABySS both those libraries. We don't usually use the FR reads, but it's possible that they could help the assembly.

That does not answer my question. Why cannot assembler try two insert-size distributions for a single input library? First trying say RF-like-insert-size, if that does not work, trying FR-like-insert-size.

Ahh, took me a loong while to realize the trimmed speaks about this 0.00005 distribution trimming. Thank you for explanation. As alway, I recommned improving the log messages and ideally also, README.md.

**Comment by mmokrejs** [comment]
> I would also like ABySS to output these stats to TSV files. I'm afraid it's not a priority as we're limited for developer time. You can scrape them out of the log files. Alternatively, it's pretty straight forward to calculate these metrics from the BAM file of reads mapped to the contigs using samtools flagstat.

No, I am not after the metrics, I am after the readnames. I haven't seen any readnames in the logfile while running in verbose mode.

> Add to your command line abyss-pe pe-bam or abyss-pe mp-bam for the contig and scaffold stages respectively, or if you prefer pe-sam and mp-sam for sam.gz files rather than BAM files.

Doh, a hidden feature to dump what abyss had in memory into SAM or BAM format? Great. Now I am starting to understand what you mean with teh soft-masked adapter. You indeed spoke about the internal BAM's. I do not have any BAM/SAM output in my `cwd` from abyss so do not be surprised I was puzzled. And I do not remeber reading anything about that in README.md either. Of course, everybody is likely to map back the original reads to the assembly, so why doesn't abyss dump out the SAM/BAM files right away? It would be even better to track errors in abyss if one had access tot he actually used alignments, and not rely on some third-party app mapping the original reads to the scaffolds.

I would gladly *add* the `abyss-pe pe-bam` and `abyss-pe mp-bam` arguments to the proposed steps but am I supposed to edit some scripts? Which? I called `abyss-pe` only once, for all steps.

**Comment by mmokrejs** [comment]
> ABySS determines the orientation of the library from these stats. As you can see, it made the correct choice in both cases.

That I understand too, that is what this Issue started with. But does really abyss need to *decide* on the orientation and insert-size? Couldn't it be more flexible? I do not say I know of another tool doing that, true.

**Comment by sjackman** [comment]
When <1% of the reads are in the other orientation, the expected gains would be marginal.

**Comment by mmokrejs** [comment]
But the reads which had no detectable linker are a mix for almost 50% RF with 50% or FR. I split the "mate-pair reads into 4 categories, the figures/logs are just for the true mate-pairs, hence so nice values. I did not dare to feed abyss with those worse categories, but provided there is so much sequence in them I really would like to. And sub-splitting these as I mentioned elsewhere, using strict, minimal assembly and a third-party mapper, is an extra manual work for me. So if abyss was willing to think of two insert-sizes&orientations per each "mate-pair" dataset, life would be easier.

**Comment by sjackman** [comment]
I recommend using both the mate-pair and unknown categories for scaffolding.
See https://github.com/sjackman/abyss-drosophila-melanogaster/blob/master/Makefile#L149
```sh
nxtrim --stdout --justmp --rf -1 foo.fq -2 foo.fq | pigz > foo.mp+unknown.fq.gz
```

**Comment by mmokrejs** [comment]
> Add to your command line abyss-pe pe-bam or abyss-pe mp-bam for the contig and scaffold stages respectively, or if you prefer pe-sam and mp-sam for sam.gz files rather than BAM files.

So I "added" the `pe-bam mp-bam` and ended up with `*.pe-3.sam.gz` and `.pe-3.bam*` files for each of my PE datafiles and with `*.-6.sam.gz` and `*-6.bam*` files for each of my MP datafiles. Obviously by specifying the "make target" to `pe-bam mp-bam` I overrrided the `default` "make target". So next time I should use `pe-bam mp-bam scaffolds scaffolds-dot stats` instead to get up to step 8, with scaffolded contigs.

```
-abyss-pe v=-v np="$threads" j="$threads" k="${kmersize}" name=tt_16D1C3L12__abyss_"${kmersize}" 
+abyss-pe v=-v np="$threads" j="$threads" k="${kmersize}" name=tt_16D1C3L12__abyss_"${kmersize}" pe-bam mp-bam scaffolds scaffolds-dot stats 
```

**Comment by sjackman** [comment]
Correct

**Comment by mmokrejs** [comment]
> Correct

make: *** No rule to make target 'scaffolds-dot'.  Stop.

Seems the following woudl be safer approach:

```
-abyss-pe v=-v np="$threads" j="$threads" k="${kmersize}" name=tt_16D1C3L12__abyss_"${kmersize}" 
+abyss-pe v=-v np="$threads" j="$threads" k="${kmersize}" name=tt_16D1C3L12__abyss_"${kmersize}" pe-bam mp-bam default 
```

**Comment by sjackman** [comment]
Ah yeah, the target is `scaffolds-graph`. I agree though using `default` is a better plan.
